File: m3ucheck.py
import requests
import re

def check_stream(url, headers):
    try:
        response = requests.get(url, headers=headers, timeout=10)
        if response.status_code == 200 and response.content:
            return 'working'
        else:
            return 'failed'
    except requests.RequestException as e:
        return f'error: {e}'

def parse_m3u(file_path):
    streams = []
    with open(file_path, 'r', encoding='utf-8') as file:
        lines = file.readlines()
        i = 0
        while i < len(lines):
            if lines[i].startswith('#EXTINF'):
                # Extract channel name
                channel_info = lines[i]
                channel_name_match = re.search(r'tvg-name="(.*?)"', channel_info)
                if channel_name_match:
                    channel_name = channel_name_match.group(1)
                else:
                    channel_name = channel_info.split(',', 1)[-1].strip()
                
                # Initialize headers
                headers = {}

                # Check for optional headers in the next few lines
                i += 1
                while lines[i].startswith('#EXTVLCOPT'):
                    if 'http-referrer' in lines[i]:
                        referrer_match = re.search(r'http-referrer=(.*)', lines[i])
                        if referrer_match:
                            headers['Referer'] = referrer_match.group(1).strip()
                    if 'user-agent' in lines[i]:
                        user_agent_match = re.search(r'user-agent=(.*)', lines[i])
                        if user_agent_match:
                            headers['User-Agent'] = user_agent_match.group(1).strip()
                    i += 1

                # Next line should be the URL
                url = lines[i].strip()
                streams.append((channel_name, url, headers))
            i += 1
    return streams

def test_streams(m3u_file, output_file):
    streams = parse_m3u(m3u_file)
    with open(output_file, 'w', encoding='utf-8') as result_file:
        for channel_name, url, headers in streams:
            status = check_stream(url, headers)
            result_file.write(f"{channel_name}: {status}\n")
            print(f"{channel_name}: {status}")

if __name__ == "__main__":
    m3u_file = "file.m3u"  # replace with the path to your M3U file
    output_file = "stream_status.txt"    # replace with the desired output file name
    test_streams(m3u_file, output_file)
    print(f"Stream status written to {output_file}")

File: m3ucheck.js
const { spawn } = require('child_process');
const fs = require('fs');
const { setTimeout } = require('timers/promises');

const m3uFilePath = 'file.m3u'; // Replace with your M3U file path
const vlcTimeout = 10000; // 10 seconds timeout for VLC to start playing the stream
const resultFilePath = 'stream_results.txt';

function parseM3U(m3uContent) {
    let lines = m3uContent.split('\n');
    let streams = [];
    let currentStream = {};

    lines.forEach(line => {
        line = line.trim();
        if (line.startsWith('#EXTINF:')) {
            let info = line.substring(8).trim();
            currentStream.name = info.split(',')[1] || ''; // Get channel name from #EXTINF
        } else if (line.startsWith('#EXTVLCOPT:http-referrer=')) {
            currentStream.referrer = line.substring(24).trim();
        } else if (line.startsWith('#EXTVLCOPT:http-user-agent=')) {
            currentStream.userAgent = line.substring(25).trim();
        } else if (line.startsWith('http')) {
            currentStream.url = line.trim();
            streams.push(currentStream);
            currentStream = {};
        }
    });

    return streams;
}

async function checkStream(stream) {
    let vlcCommand = `vlc --intf dummy --no-video --play-and-exit "${stream.url}"`;

    let vlcProcess = spawn(vlcCommand, [], { shell: true });

    let timeoutPromise = setTimeout(vlcTimeout);

    let hasError = false;

    vlcProcess.stdout.on('data', (data) => {
        let output = data.toString();
        if (output.includes('main input error')) {
            hasError = true;
        }
    });

    vlcProcess.stderr.on('data', (data) => {
        let output = data.toString();
        if (output.includes('cannot open') || output.includes('main input error')) {
            hasError = true;
        }
    });

    try {
        await Promise.race([timeoutPromise]);

        if (!hasError) {
            console.log(`${stream.name} - working`);
            appendResult(`${stream.name} - working`);
        } else {
            console.log(`${stream.name} - failed to start playback`);
            appendResult(`${stream.name} - failed to start playback`);
        }
    } catch (err) {
        console.error(`${stream.name} - failed (error: ${err.message})`);
        appendResult(`${stream.name} - failed (error: ${err.message})`);
    } finally {
        vlcProcess.kill();
    }
}

function appendResult(result) {
    fs.appendFileSync(resultFilePath, result + '\n', (err) => {
        if (err) {
            console.error('Error appending to stream_results.txt:', err);
        }
    });
}

// Check if the result file exists and delete it if it does
fs.access(resultFilePath, fs.constants.F_OK, (err) => {
    if (!err) {
        // File exists, so delete it
        fs.unlink(resultFilePath, (err) => {
            if (err) {
                console.error('Error deleting existing stream_results.txt:', err);
            }
        });
    }
});

fs.readFile(m3uFilePath, 'utf8', async (err, data) => {
    if (err) {
        console.error(`Error reading file: ${err}`);
        return;
    }

    const streams = parseM3U(data);

    for (let stream of streams) {
        if (!stream.name) {
            stream.name = stream.url.split('/').pop().split('.')[0]; // Use filename as fallback name
        }
        await checkStream(stream);
    }

    console.log('All streams checked.');

    // Close VLC or perform cleanup here if needed
    process.exit(0); // Ensure script terminates after all streams are checked
});
